pyomo/contrib/doe/measurements.py

```diff
@@ -25,252 +25,270 @@
         """
-        measurement_name = list(all_info.keys())
-        # a list of measurement extra indexes
-        measurement_extra_index = []
-        # check if the measurement has extra indexes
-        for i in measurement_name:
-            if type(all_info[i]) is dict:
-                index_list = list(all_info[i].keys())
-                measurement_extra_index.append(index_list)
-            elif type(all_info[i]) is list:
-                measurement_extra_index.append(None)
-        # a dictionary, keys are measurement names, values are a list of extra indexes
-        self.name_and_index = dict(zip(measurement_name, measurement_extra_index))
-
-    def _generate_flatten_name(self, measure_name_and_index):
+        self.special_set_value = self._generate_dict(values)
+        
+
+    def _generate_dict(self, values):
         """
-        Generate measurement flattened names
-        Parameters
-        ----------
-        measure_name_and_index: a dictionary, keys are measurement names, values are lists of extra indexes
+        Given a list of values, return a dictionary, keys are special set names. 
+        """
+        value_map = {}
+        for i in range(len(values)):
+            value_map[self.special_set[i]] = values[i]
 
-        Returns
-        ------
-        jac_involved_name: a list of flattened measurement names
+        return value_map
+
+
+    def _add_elements(self, var_name, extra_index=None, time_index=None):
         """
-        flatten_names = []
-        for j in measure_name_and_index.keys():
-            if measure_name_and_index[j] is not None: # if it has extra index
-                for ind in measure_name_and_index[j]:
-                    flatten_name = j + self.ind_string + str(ind)
-                    flatten_names.append(flatten_name)
-            else:
-                flatten_names.append(j)
-
-        self.flatten_measure_name = flatten_names
-
-    def _generate_variance(self, flatten_measure_name, variance, name_and_index):
+        Used for generating string names with indexes. 
+
+        Parameter 
+        ---------
+        var_name: a ``list`` of var names 
+        extra_index: a ``list`` containing extra indexes except for time indexes 
+            if default (None), no extra indexes needed for all var in var_name
+            if it is a nested list, it is a ``list`` of ``list`` of ``list``, 
+            they are different multiple sets of indexes for different var_name
+            for e.g., extra_index[0] are all indexes for var_name[0], extra_index[0][0] are the first index for var_name[0]
+        time_index: a ``list`` containing time indexes
+            default choice is None, means this is a model parameter 
+            if it is an algebraic variable, time index should be set up to [0]
+            if it is a nested list, it is a ``list`` of ``lists``, they are different time set for different var in var_name
```

Comment:
In the doc string, we should provide an example input such as  "extra_index = [CA, CB, CC], time_index=[0,1,2]".

Then throughout the function, you can write more clear comments the explain what is happening with this example.